slatedb/src/dispatcher.rs

```diff
@@ -0,0 +1,813 @@
+    ///
+    /// ## Arguments
+    ///
+    /// * `handler`: The [MessageHandler] to use for processing messages.
+    /// * `rx`: The [mpsc::UnboundedReceiver<T>] to use for receiving messages.
+    /// * `clock`: The [SystemClock] to use for time.
+    /// * `cancellation_token`: The [CancellationToken] to use for shutdown.
+    /// * `state`: The [DbState] to use for error tracking. If provided, the dispatcher
+    ///   will set [DbState::error] when an error is encountered.
+    pub(crate) fn new_with_fp_registry(
+        handler: Box<dyn MessageHandler<T>>,
+        rx: mpsc::UnboundedReceiver<T>,
+        clock: Arc<dyn SystemClock>,
+        cancellation_token: CancellationToken,
+        error_state: WatchableOnceCell<SlateDBError>,
+        fp_registry: Arc<FailPointRegistry>,
+    ) -> Self {
+        Self {
+            handler,
+            rx,
+            clock,
+            cancellation_token,
+            error_state,
+            fp_registry,
+        }
+    }
+
+    /// Runs the dispatcher. This is the primary entry point for running the dispatcher.
+    ///
+    /// [MessageDispatcher::run] is the primary entry point for running the dispatcher;
+    /// it is responsible for:
+    ///
+    /// 1. Running the main event loop ([MessageDispatcher::run_loop]).
+    /// 2. Processing the final result when the event loop exits
+    ///    ([MessageDispatcher::handle_result]).
+    /// 4. Cleaning up resources and processing any remaining messages
+    ///    ([MessageDispatcher::cleanup]).
+    ///
+    /// ## Returns
+    ///
+    /// A [Result] containing the [DbState::error] if it was already set, or the result
+    /// of [MessageDispatcher::handle_result] otherwise.
+    pub(crate) async fn run(&mut self) -> Result<(), SlateDBError> {
+        // TODO: handle panic here (similar to spawn_bg_task)
+        let result = self.run_loop().await;
+        let result = self.handle_result(result);
+        if let Err(e) = self.cleanup(result.clone()).await {
+            warn!("failed to cleanup dispatcher on shutdown [error={:?}]", e);
+        }
+        result
+    }
+
+    /// Runs the main event loop for the dispatcher. This is where messages and ticker
+    /// events are processed.
+    ///
+    /// [MessageDispatcher::run_loop] contains a message loop with the following control
+    /// flow:
+    ///
+    /// 1. Break immediately if we're in an error state or cancelled, and return
+    ///    [SlateDBError::BackgroundTaskShutdown].
+    /// 2. Else, if there is a message, read it and invoke [MessageHandler::handle].
+    /// 3. Else, if there is a ticker event, read it and invoke [MessageHandler::handle].
+    ///
+    /// If there is an uncaught error at any point, the error is returned immediately (in
+    /// lieu of [SlateDBError::BackgroundTaskShutdown] on a clean shutdown).
+    ///
+    /// ## Returns
+    ///
+    /// A [Result] containing [SlateDBError::BackgroundTaskShutdown] on clean shutdown,
```

Comment:
(can be in a future PR if we agree to change it) this has always confused me a bit - any reason not to make it `Ok(BackgroundTaskShutdown)` or something like that? Why is this an "error"?

Reply:
Opened: https://github.com/slatedb/slatedb/issues/817